internal/fields/validate.go

```diff
@@ -345,18 +370,109 @@ func (v *Validator) validateScalarElement(key string, val interface{}, doc commo
+		if err != nil {
+			return fmt.Errorf("field %q is not normalized as expected: %w", key, err)
+		}
 	}
 
-	err = v.parseElementValue(key, *definition, val, doc)
+	err := v.parseElementValue(key, *definition, val, doc)
 	if err != nil {
-		return errors.Wrap(err, "parsing field value failed")
+		return fmt.Errorf("parsing field value failed: %w", err)
 	}
 	return nil
 }
 
+func (v *Validator) SanitizeSyntheticSourceDocs(docs []common.MapStr) ([]common.MapStr, error) {
+	var newDocs []common.MapStr
+	for _, doc := range docs {
+		for key, contents := range doc {
+			definition := FindElementDefinition(key, v.Schema)
+			if definition == nil {
+				continue
+			}
+
+			if !v.disabledDependencyManagement && definition.External != "" {
+				def, err := v.FieldDependencyManager.ImportField(definition.External, key)
+				if err != nil {
+					return nil, fmt.Errorf("can't import field (field: %s): %w", key, err)
+				}
+				definition = &def
+			}
+
+			shouldBeArray := false
+			// normalization should just be checked if synthetic source is enabled and the
+			// spec version of this package is >= 2.0.0
+			if v.disabledNormalization && !v.specVersion.LessThan(semver2_0_0) {
+				for _, normalize := range definition.Normalize {
+					switch normalize {
+					case "array":
+						shouldBeArray = true
+					}
+					if shouldBeArray {
+						break
+					}
+				}
+			}
+			// if it needs to be normalized, the field is kept as it is
+			if shouldBeArray {
+				continue
+			}
+			// in case it is not specified any normalization and that field is an array of
+			// just one element, the field is going to be updated to remove the array and keep
+			// that element as a value.
+			vals, ok := contents.([]interface{})
+			if !ok {
+				continue
+			}
+			if len(vals) == 1 {
+				_, err := doc.Put(key, vals[0])
+				if err != nil {
+					return nil, fmt.Errorf("key %s could not be updated: %w", key, err)
+				}
+			}
+		}
+		expandedDoc, err := createDocExpandingObjects(doc)
+		if err != nil {
+			return nil, fmt.Errorf("failure while expanding objects from doc: %w", err)
+		}
+
+		newDocs = append(newDocs, expandedDoc)
+	}
+	return newDocs, nil
+}
+
+func createDocExpandingObjects(doc common.MapStr) (common.MapStr, error) {
+	keys := make([]string, 0)
+	for k, _ := range doc {
+		keys = append(keys, k)
+	}
+	sort.Strings(keys)
+
+	newDoc := make(common.MapStr)
+	for _, k := range keys {
+		value, err := doc.GetValue(k)
+		if err != nil {
+			return nil, fmt.Errorf("not found key %s: %w", k, err)
+		}
+
+		_, err = newDoc.Put(k, value)
+		if err == nil {
+			continue
+		}
+
+		// Possible errors found but not limited to those
+		// - expected map but type is string
+		// - expected map but type is []interface{}
+		if strings.HasPrefix(err.Error(), "expected map but type is") {
+			logger.Warnf("not able to add key %s, is this a multifield?: %s", k, err)
+			continue
+		}
+		return nil, fmt.Errorf("not added key %s with value %s: %w", k, value, err)
```

Comment:
Examples found in my local tests:

```
2023/06/07 10:41:26  WARN not able to add key host.os.name.text, is this a multifield?: expected map but type is string
2023/06/07 10:41:26  WARN not able to add key metricset.name.text, is this a multifield?: expected map but type is []interface {}
```

As there could be other keys for the multifield a part from `text` I didn't want to hardcode that

Reply:
I would say that these errors about "expected map" are expected with multi-fields, I think we can log them at the debug level, or don't log anything, but don't log them at the warn level.